pyodide_lock/spec.py

```diff
@@ -91,3 +101,151 @@ def check_wheel_filenames(self) -> None:
+        # n.b. this assumes existing packages have correct dependencies,
+        # which is probably a good assumption.
+
+        requirements_with_extras = []
+        for package in new_packages.values():
+            # add any requirements to the list of packages
+            our_depends = []
+            wheel_file = new_package_wheels[package.name]
+            requirements = get_wheel_dependencies(wheel_file, package.name)
+            for r in requirements:
+                req_marker = r.marker
+                req_name = canonicalize_name(r.name)
+                if req_marker is not None:
+                    if not req_marker.evaluate(
+                        {"sys_platform": "emscripten", "platform_system": "Emscripten"}
+                    ):
+                        # not used in pyodide / emscripten
+                        # or optional requirement
+                        continue
+                if r.extras:
+                    # this requirement has some extras, we need to check
+                    # that the required package depends on these extras also.
+                    requirements_with_extras.append(r)
+                if req_name in new_packages or req_name in self.packages:
+                    our_depends.append(req_name)
+                else:
+                    raise RuntimeError(
+                        f"Requirement {req_name} is not in this distribution."
+                    )
+            package.depends = our_depends
+
+        while len(requirements_with_extras) != 0:
+            extra_req = requirements_with_extras.pop()
+            extra_package_name = canonicalize_name(r.name)
+            if extra_package_name not in new_packages:
+                continue
+            package = new_packages[extra_package_name]
+            our_depends = package.depends
+            wheel_file = new_package_wheels[package.name]
+            requirements = get_wheel_dependencies(wheel_file, package.name)
+            for extra in extra_req.extras:
+                for r in requirements:
+                    req_marker = r.marker
+                    req_name = canonicalize_name(r.name)
+                    if req_marker is None:
+                        continue
+                    if req_marker.evaluate(
+                        {
+                            "sys_platform": "emscripten",
+                            "platform_system": "Emscripten",
+                            "extra": extra,
+                        }
+                    ):
+                        if req_name in new_packages or req_name in self.packages:
+                            our_depends.append(req_name)
+                            if r.extras:
+                                requirements_with_extras.append(r)
+                        else:
+                            raise RuntimeError(
+                                f"Requirement {req_name} is not in this distribution."
+                            )
```

Comment:
It's a very long method. Some of it should probably be split into a separate helper function as @bollwyvl  did in https://github.com/pyodide/pyodide-lock/pull/18